tools/python/boutiques/puller.py

```diff
@@ -15,43 +16,65 @@
-        self.verbose = verbose
-        self.sandbox = sandbox
+        self.zenodo_entries = []
         self.cache_dir = os.path.join(os.path.expanduser('~'), ".cache",
                                       "boutiques")
-        self.cached_fname = os.path.join(self.cache_dir,
-                                         "zenodo-{0}.json".format(self.zid))
+        discarded_zids = zids
+        # This removes duplicates, should maintain order
+        zids = list(dict.fromkeys(zids))
+        for zid in zids:
+            discarded_zids.remove(zid)
+            try:
+                newzid = zid.split(".", 1)[1]
+                newfname = os.path.join(self.cache_dir,
+                                        "zenodo-{0}.json".format(newzid))
+                self.zenodo_entries.append({"zid": newzid, "fname": newfname})
+            except IndexError:
+                raise_error(ZenodoError, "Zenodo ID must be prefixed by "
+                            "'zenodo', e.g. zenodo.123456")
+        self.verbose = verbose
+        self.sandbox = sandbox
+        if(self.verbose):
+            for zid in discarded_zids:
+                print_info("Discarded duplicate id {0}".format(zid))
 
     def pull(self):
         # return cached file if it exists
-        if os.path.isfile(self.cached_fname):
-            if(self.verbose):
-                print_info("Found cached file at %s"
-                           % self.cached_fname)
-            return self.cached_fname
+        json_files = []
+        for entry in self.zenodo_entries:
+            if os.path.isfile(entry["fname"]):
+                if(self.verbose):
+                    print_info("Found cached file at %s"
+                               % entry["fname"])
+                json_files.append(entry["fname"])
+                continue
 
-        from boutiques.searcher import Searcher
-        searcher = Searcher(self.zid, self.verbose, self.sandbox,
-                            exact_match=True)
-        r = searcher.zenodo_search()
+            searcher = Searcher(entry["zid"], self.verbose, self.sandbox,
+                                exact_match=True)
+            r = searcher.zenodo_search()
 
-        for hit in r.json()["hits"]["hits"]:
-            file_path = hit["files"][0]["links"]["self"]
-            file_name = file_path.split(os.sep)[-1]
-            if hit["id"] == int(self.zid):
-                if not os.path.exists(self.cache_dir):
-                    os.makedirs(self.cache_dir)
-                if(self.verbose):
-                    print_info("Downloading descriptor %s"
-                               % file_name)
-                downloaded = urlretrieve(file_path, self.cached_fname)
-                print("Downloaded descriptor to " + downloaded[0])
-                return downloaded[0]
-        raise_error(ZenodoError, "Descriptor not found")
+            if not len(r.json()["hits"]["hits"]):
+                raise_error(ZenodoError, "Descriptor \"{0}\" "
```

Comment:
I think it might be nicer if, instead of raising an exception after the first invalid Zenodo ID, we simply printed an error message without terminating the program. That way, the valid ones would still be downloaded and the user would only have to rerun the command for the invalid ones. The logger has a `print_error` function for this purpose.